kolibri/core/analytics/utils.py

```diff
@@ -62,6 +75,128 @@
 ]
 
 
+def standard_deviation(data):
+    if data:
+        # calculate mean
+        m = sum(data) / len(data)
+        # return population std
+        return round(math.sqrt(sum((d - m) ** 2 for d in data) / len(data)), 2)
+    return None
+
+
+def calculate_demographic_stats(dataset_id=None, channel_id=None, learner=True):
+    stats = {}
+    roles_filter = Q(roles__isnull=False)
+    if learner:
+        roles_filter = Q(roles__isnull=True)
+    queryset = FacilityUser.objects.filter(roles_filter)
+    # handle stats at facility level
+    if dataset_id:
+        queryset = queryset.filter(dataset_id=dataset_id)
+    # handle stats at channel level
+    if channel_id:
+        user_ids = (
+            queryset.filter(contentsummarylog__channel_id=channel_id)
+            .distinct()
+            .values_list("id", flat=True)
+        )
+        # pass distinct user_ids as subquery
+        queryset = FacilityUser.objects.filter(id__in=user_ids)
+
+    # calculate stats if there are USER_THRESHOLD users or more
+    if queryset.count() >= USER_THRESHOLD:
+        # get list of all birth years to calculate variance
+        list_of_birth_years = (
+            queryset.exclude(
+                Q(birth_year="")
+                | Q(birth_year=demographics.NOT_SPECIFIED)
+                | Q(birth_year=demographics.DEFERRED)
+            )
```

Comment:
Could be interesting to benchmark (at some point) against doing `birth_year__in=["", ...]` instead.

Reply:
Result: same speed, but a bit less code.